src/Storages/DataLakes/IcebergMetadataParser.cpp

```diff
@@ -226,59 +223,100 @@ struct IcebergMetadataParser<Configuration, MetadataReadHelper>::Impl
+            }
+
+            if (columns.at(0)->getDataType() != TypeIndex::Int32)
+            {
+                throw Exception(ErrorCodes::ILLEGAL_COLUMN,
+                                "The parsed column from Avro file of `status` field should be Int32 type, got {}",
+                                columns.at(0)->getFamilyName());
+            }
+            if (columns.at(1)->getDataType() != TypeIndex::Tuple)
+            {
+                throw Exception(ErrorCodes::ILLEGAL_COLUMN,
+                                "The parsed column from Avro file of `file_path` field should be Tuple type, got {}",
+                                columns.at(1)->getFamilyName());
+            }
+
+            const auto status_int_column = assert_cast<ColumnInt32 *>(columns.at(0).get());
+            const auto data_file_tuple_column = assert_cast<ColumnTuple *>(columns.at(1).get());
+
+            if (status_int_column->size() != data_file_tuple_column->size())
+            {
+                throw Exception(ErrorCodes::ILLEGAL_COLUMN,
+                                "The parsed column from Avro file of `file_path` and `status` have different rows number: {} and {}",
+                                status_int_column->size(), data_file_tuple_column->size());
             }
 
-            const auto * str_col = assert_cast<const ColumnString *>(col_str.get());
-            for (size_t i = 0; i < str_col->size(); ++i)
+            const auto * data_file_name_column = metadata.format_version == 1
+                ? data_file_tuple_column->getColumnPtr(0).get()
+                : data_file_tuple_column->getColumnPtr(1).get();
+
+            if (data_file_name_column->getDataType() != TypeIndex::String)
+            {
+                throw Exception(ErrorCodes::ILLEGAL_COLUMN,
+                                "The parsed column from Avro file of `file_path` field should be String type, got {}",
+                                data_file_name_column->getFamilyName());
+            }
+            auto file_name_str_column = assert_cast<const ColumnString *>(data_file_name_column);
+
+            for (size_t i = 0; i < status_int_column->size(); ++i)
             {
-                const auto data_path = std::string(str_col->getDataAt(i).toView());
+                const auto status = status_int_column->getInt(i);
+                const auto data_path = std::string(file_name_str_column->getDataAt(i).toView());
                 const auto pos = data_path.find(configuration.url.key);
+                const auto file_path = data_path.substr(pos);
                 if (pos == std::string::npos)
                     throw Exception(ErrorCodes::BAD_ARGUMENTS, "Expected to find {} in data path: {}", configuration.url.key, data_path);
-                keys.emplace_back(data_path.substr(pos));
+
+                if (status == 2)
+                    keys.erase(file_path);
```

Comment:
In what case this path will already be in `keys`? I think if file was deleted, it should be just in one manifest file and with only `status = 2`